harness/determined/searcher/_search_runner.py

```diff
@@ -0,0 +1,351 @@
+            )
+            request_id = uuid.UUID(event.trialProgress.requestId)
+            self.state.trial_progress[request_id] = float(event.trialProgress.partialUnits)
+            progress = self.search_method.progress(self.state)
+            operations = [searcher.Progress(progress)]
+        else:
+            raise RuntimeError(f"Unsupported event {event}")
+        return operations
+
+    def run_experiment(
+        self,
+        experiment_id: int,
+        session: client.Session,
+        prior_operations: Optional[List[searcher.Operation]],
+    ) -> None:
+        experiment_is_active = True
+
+        try:
+            while experiment_is_active:
+                time.sleep(
+                    2
+                )  # we don't want to call long polling API more often than every 2 seconds.
+                events = self.get_events(session, experiment_id)
+                if not events:
+                    continue
+                logger.info(json.dumps([SearchRunner._searcher_event_as_dict(e) for e in events]))
+                # the first event is an event we have already processed and told master about it
+                # however, we may not have saved the state after that event if we crashed
+                # after POSTing operations but before saving state
+                last_event_id = self.state.last_event_id
+                first_event = True
+                for event in events:
+                    if (
+                        first_event
+                        and last_event_id != 0
+                        and last_event_id >= event.id >= 0
+                        and prior_operations is not None
+                    ):
+                        logger.info(f"Resubmitting operations for event.id={event.id}")
+                        operations = prior_operations
+                    else:
+                        if event.experimentInactive:
+                            logger.info(
+                                f"experiment {self.state.experiment_id} is "
+                                f"inactive; state={event.experimentInactive.experimentState}"
+                            )
+                            if (
+                                event.experimentInactive.experimentState
+                                == bindings.determinedexperimentv1State.STATE_COMPLETED
+                            ):
+                                self.state.experiment_completed = True
+
+                            if (
+                                event.experimentInactive.experimentState
+                                == bindings.determinedexperimentv1State.STATE_PAUSED
+                            ):
+                                self._show_experiment_paused_msg()
+                            else:
+                                experiment_is_active = False
+                            break
+
+                        else:
+                            operations = self._get_operations(event)
```

Comment:
small nit: since the previous if statement ends in a break, eliminate the `else` and unindent `operations = ...`

It's confusing that:
```
operations = self._get_operations(event)
```
and
```
# save state
self.state.last_event_id = event.id
self.save_state(experiment_id, operations)
```
appear at different levels of indentation, but are logically in the same block of code that always runs as a unit.

Reply:
Wait, that would completely alter the logic!

Reply:
👍